### DISCUSSION TITLE
Systemvariablen und Programme in HA - Verbesserungsvorschlag

### DISCUSSION THREAD

**Discussion by Baxxy13** [SukramJ/aiohomematic#1923] [Ideas]
Ich schreibe in folgendem nur über die SysVars, das lässt sich aber auch direkt auf die Programme übertragen.

Stand heute hat man 2 Optionen. Entweder alle SysVars in HA oder keine. 
Nur ein oder zwei ausgewählte SysVars geht aktuell nicht. Ist jetzt nicht das große Problem da ja erstmal alle SysVars nur als deaktivierte Entitäten vorhanden sind. Aber auch die deaktivierten sind ja irgendwie in HA "registriert" und das gefällt mir nicht so recht.

**Folgende** Idee, abgeguckt beim CCU-Jack:
- nur in der WebUI (CCU/RM) "markierte" SysVars werden eingelesen und als Entitäten angelegt

**Grobe Umsetzung:**
- beim Start der Integration wird einmal über alle SysVars iteriert und die ID's der "markierten" in ein temporäres ID-Array abgelegt. Zusätzlich werden für diese "markierten" SysVars die Entitäten erzeugt
- das SysVar-Polling nutzt nun das ID-Array um damit über die "markierten" SysVars zu iterieren um deren aktuellen Status zu lesen.
- ein zusätzlicher zyklischer Service nutzt die gleiche Funktion wie beim Start der Integration um das ID-Array "aufzufrischen" falls es neue oder gelöschte "markierte" SysVars gibt 

**Pro's:**
- man hat nur die SysVars als Entitäten die man explizit ausgewählt hat
- das SysVar-Polling fragt nicht alle SysVars ab, sondern nur die ausgewählten (hier weiß ich nicht wie das aktuell umgesetzt ist)
- neu "markierte" SysVars werden (nach dem zykl. "Refresh-Polling") automatisch zu HA hinzugefügt bzw. "demarkierte" werden "nicht verfügbar" (weiß nicht ob das so ginge)
- man könnte die SysVars gleich als aktive Entitäten bereitstellen weil man ja "weiß" das der User diese haben möchte weil er sie markiert hat. 

**Contra's:**
- der User muss aktiv die SysVars in der WebUI "markieren"
- es bräuchte einen zweiten Marker für die SysVars, der Marker `hahm` für die EXTENDED - SysVars existiert ja schon es bräuchte aber noch einen für die DEFAULT - SysVars
- Programme bräuchten einen Marker 

Falls die Idee gefällt kann ich gerne bei den Scripten helfen, vom Rest verstehe ich leider nicht genug.

**Comment by SukramJ** [comment]
Mir gefällt die Idee.

- In der Konfiguration könnte man die Marker angeben, die man geladen haben möchte, ansonsten werden weiterhin alle Sysvars geladen. 
- Markierte Sysvars werden initial als aktiv angelegt, dadurch bleiben diese im Lifecycle, und können auch durch die Integration gelöscht werden.
- `hahm` wird immer geladen
- Die Abfrage an die CCU kann gleich bleiben, aber es werden nur markierte weiterverarbeitet.

Cooler wäre natürlich wenn in RM dazu das Beschreibungsfeld nicht verwendet werden müsste, sondern es eine zusätzliche Spalte mit Markern/Label/Tags gäbe.

**Reply by SukramJ** [reply]
Solllen die Marker danach aus der Beschreibung (Attribut in HA) entfernt werden? Aktuell passiert das nur für `hahm`.

**Reply by Baxxy13** [reply]
Könnte man machen, weiß nicht wie wichtig das den "SysVar-in-HA" - Nutzern wäre.

**Comment by Baxxy13** [comment]
Mir ist leider noch etwas Contra eingefallen.
- Es gibt ja SysVars an die kommt man speziell auf nicht RaspberryMatics "schlecht ran". (Alarmmeldungen / Servicemeldungen)
Die Frage wäre hier ob man die beiden immer einliest. (außer der SysVar-Scan ist deaktiviert)

- Dann haben wir noch die systemgenerierten Watchdog SysVars (nur auf RM) die sich auch nicht markern lassen.

**Reply by SukramJ** [reply]
sind das die `intern` markierten?

**Reply by Baxxy13** [reply]
Ja, `Alarmmeldungen` und `Servicemeldungen` sind per default `internal = true` und `visible = false`.
Ist mit RM kein Problem weil es ja einen Button zum Einblenden gibt.
Den gibt's bei allen anderen Systemen nicht so das man entweder:
- die Beschreibung per Script setzen muss (für den Marker)
oder
- `internal = false` per Script setzen muss um die anschließend über die WebUI zu "markieren"

Die RM-Watchdog-Alarme sind noch "komplizierter", hier wird bei jeder Auslösung `.DPInfo()` neu geschrieben.
Man kann also weder einen Marker in den Namen setzen (es würde einfach eine neue SV ohne den Marker im Namen erzeugt) noch kann man die Beschreibung für den Marker nutzen.

**Reply by SukramJ** [reply]
~~Also eine Checkbox um interne SysVars zu laden (default=True). Diese sind beim Erstellen deaktiviert.~~
Edit: Interne Werte werden immer geladen. Diese sind beim Erstellen deaktiviert.
und
bei nicht internen Sysvars:
- eine kommaseparierte Liste mit Markern, die die zu verarbeitenden Werte einschränkt. Diese sind beim Erstellen aktiviert.
- ist die kommaseparierte Liste leer, dann werden alle Werte verarbeitet. Diese sind beim Erstellen deaktiviert.

**Reply by Baxxy13** [reply]
Klingt gut für mich.

**Reply by SukramJ** [reply]
Versuch mal die 1.76.0b0.

Die UI Eingabe (unter erweiterte Optionen) muss ich noch optimieren, aber grundsätzlich sollte es funktionieren.

**Reply by Baxxy13** [reply]
Also der `hahm` Marker für die EXTENDED - SysVars bleibt bestehen, den muss ich nicht in die Liste eintragen?
Selbst definierbare Marker finde ich echt praktisch, wobei ja eigentlich einer reichen würde. Aber mehr ist immer gut.

Der erste Testdurchlauf mit jeweils 2 verschiedenen Markern für SysVars und Programme sah sehr gut aus.

Was mir aber aufgefallen ist:
**SysVar:**
- Steht der Marker als einziges in der Beschreibung oder direkt am Anfang (bspw: `SV-HA rssi Durchschnitt` klappt alles
- steht der Marker am Ende der Beschreibung (bspw: `rssi Durchschnitt SV-HA` ) wird er wohl nicht erkannt

**Programme:**
- die werden aktuell nicht aus HA gelöscht wenn ich in der WebUI den Marker entferne, denke da hast du noch keinen zyklischen Service für
- Die "gemarkerten" Programme würde ich auch gleich als aktive Entität anlegen. Es steht ja quasi fest das der User die in HA "nutzen" möchte weil er sie explizit markiert hat.

Ich finde du hast die Idee sehr gut umgesetzt. 
Keine unerwünschten deaktivierten Entitäten und um das Löschen nicht mehr benötigter SysVars braucht man sich nun auch nicht mehr kümmern. Das war ja immer ein ziemlicher Krampf.

Auf jeden Fall vielen Dank dafür!

**Reply by SukramJ** [reply]
Danke für die Rückmeldung, eine neue Version folgt bald.

Ich würde gerne 4 feste Marker verwenden. Das macht die UI einfacher:

    HA = normale Variablen/Programme
    HAHM =  erweiterte Variablen
    INTERNAL = Interne variable/Programme
    MQTT = Wenn diese schon für CCU-Jack getaggt sind

Spricht was dagegen?

**Reply by Baxxy13** [reply]
Ich denke nicht. 
Mir wäre nur wichtig das die Marker auch am Ende der Beschreibung stehen können.

Bspw. `dies ist eine Test-SysVar HA`

Und möglichst keine false positives, also bspw. `HAch ist die SysVar praktisch`

**Reply by SukramJ** [reply]
> HAch ist die SysVar praktisch

Der Fall wäre mir eigentlich egal.

Hast du eine kurze alternative für `HA`?

**Reply by SukramJ** [reply]
Aktuell vergleiche ich _case sensitive_.
`HA`, `hahm`, `INTERNAL`, `MQTT`

`hahm` ist klein geschrieben weil alt.

**Reply by SukramJ** [reply]
Ich mach aus dem HA ein HX. Damit gibt es weniger Kollisionen.

**Reply by Baxxy13** [reply]
HX gefällt mir. Kommt ja normalerweise nicht vor.
Wofür hast du jetzt den INTERNAL? 
Du hattest doch geschrieben das die internen immer geladen und als deaktivierte angelegt werden.

**Reply by SukramJ** [reply]
Ich hab das noch mal geändert.
INTERNAL, damit die internen Werte geladen werden. In der CCU muss dazu nichts in die Beschreibung geschrieben werden, sondern es wird das `intern` flag ausgewertet.

**Reply by SukramJ** [reply]
Die 1.76.0b1 steht bereit. 
Die Integration muss noch mal konfiguriert werden.
Siehe auch https://github.com/SukramJ/custom_homematic?tab=readme-ov-file#filtering-system-variables-and-programs

**Reply by Baxxy13** [reply]
Super, gucke ich mir später an. 

Ich hätte nen Breaking-Change gemacht und `hahm ` zu `HAHM` oder besser `HXHM` geändert. (das X könnte man als Xtended interpretieren)

Warum?:
- ich mag es gerne einheitlich
- groß geschriebene Marker sind leichter in der WebUI zu erkennen

Aber ich weiß, Breaking Changes sind immer Mist. 😉

**Reply by SukramJ** [reply]
Wird mal wieder Zeit für einen Breaking Change.

Kann Du ein Skript schreiben, das die Beschreibungen von hahm auf HAHM ändert?

**Reply by Baxxy13** [reply]
Klar. Mache ich später fertig.

**Reply by SukramJ** [reply]
Ich hab mich noch mal gegen HXHM und für HAHM entschieden, da HX nicht in HAHM enthalten ist.

**Reply by Baxxy13** [reply]
So, das hier sollte es in einem Rutsch erledigen.

```
!- Suchen und Ersetzen in SysVar-Beschreibungen
string sSuchen   = "hahm";
string sErsetzen = "HAHM";

string sSV_ID;
foreach(sSV_ID, dom.GetObject(ID_SYSTEM_VARIABLES).EnumIDs()) {
  object oSV = dom.GetObject(sSV_ID);
  string sBeschreibung = oSV.DPInfo();
  if (sBeschreibung.Contains(sSuchen)) {
    sBeschreibung = sBeschreibung.Replace(sSuchen, sErsetzen);
    oSV.DPInfo(sBeschreibung);
    WriteLine(oSV.Name() #" | "# sSuchen #" ersetzt durch: "# sErsetzen);
  }
}
```

**Reply by SukramJ** [reply]
1.76.0b2 verwendet jetzt HAHM

**Reply by Baxxy13** [reply]
Also mit den SysVars klappt jetzt alles wie erwartet. 
Sie werden hinzugefügt wenn ich sie in der WebUI markiere und entfernt wenn ich den Marker entferne.
(also die HX und HAHM markierten, die internen habe ich jetzt erstmal nicht weiter gefprüft)

Programme werden auch nach dem markieren in der WebUI hinzugefügt, aber sie werden nicht entfernt wenn ich den Marker entferne.
Hier braucht's dann einen HA-Neustart, dann werden die Programme als "nicht mehr bereitgestellt" gelistet und ich kann sie händisch löschen.

**Reply by SukramJ** [reply]
Bei mir funktioniert es auch mit den Programmen. 
Welchen Marker verwendest du?

**Reply by Baxxy13** [reply]
So eingestellt und HX als Marker für's Programm verwendet.
![Integration_Markers](https://github.com/user-attachments/assets/c9627921-d54f-47ce-9eb4-b77f4f0f04fe)

Hier nochmal als gif, das Programm bleibt hier auch nach Stunden noch aktiv.
![GIF 22 12 2024 09-36-26](https://github.com/user-attachments/assets/a3e87a79-e87c-4537-914e-cfc411681371)

**Reply by SukramJ** [reply]
So mach ich das auch, aber es funktioniert.

Der Mechanismus zum entfernen der Programme ist übrigens schon alt, den hab ich nicht jetzt erst eingebaut.

**Reply by Baxxy13** [reply]
Hmm, liegt es vielleicht an meiner "Multi-Integration"? 
Ich habe aktuell 3 Zentralen in einen HA integriert, primär zu Testzwecken.
Ich versuche das mit dem Programm mal mit einer der anderen beiden integrierten Zentralen.

**Reply by SukramJ** [reply]
> Hmm, liegt es vielleicht an meiner "Multi-Integration"?

Ich hab bei mir immer zwei Instanzen der Integration aktiv, und es funktioniert auf Beiden.

... und auch auf mehreren HA Instanzen.

**Reply by Baxxy13** [reply]
Bei mir leider nicht. 
Wie debugge ich das am Besten? 
Mittels Debug-Logging bei der Integration oder spezielle Logger-Einstellungen in der configuration.yaml?

**Reply by SukramJ** [reply]
Ich weiss nicht wie hilfreich die debug Infos bei dem Thema sind, aber versuch mal:

```
logger:
  default: warn
  logs:
    hahomematic: debug
    custom_components.homematicip_local: debug
```

**Reply by Baxxy13** [reply]
Hmm, ist in der Tat nur bedingt aussagekräftig.

Nach dem markieren in der WebUI:
```
Dec 22 11:03:08 HA-Test-Pi4B2GB-45 homeassistant[605]: 2024-12-22 12:03:08.183 DEBUG (MainThread) [hahomematic.client.json_rpc] GET_PROGRAM_DESCRIPTIONS: Getting program descriptions
Dec 22 11:03:08 HA-Test-Pi4B2GB-45 homeassistant[605]: 2024-12-22 12:03:08.184 DEBUG (MainThread) [hahomematic.model.hub] UPDATE_PROGRAM_DATA_POINTS: 1 programs received for RM80-CM4-4GB
```

Nach dem Löschen des Markers:
```
Dec 22 11:04:29 HA-Test-Pi4B2GB-45 homeassistant[605]: 2024-12-22 12:04:29.533 DEBUG (MainThread) [hahomematic.client.json_rpc] GET_PROGRAM_DESCRIPTIONS: Getting program descriptions
Dec 22 11:04:29 HA-Test-Pi4B2GB-45 homeassistant[605]: 2024-12-22 12:04:29.534 DEBUG (MainThread) [hahomematic.model.hub] UPDATE_PROGRAM_DATA_POINTS: No programs received for RM80-CM4-4GB
```

Die Programm-Entität bleibt aber bestehen und lässt sich auch weiterhin bedienen. (das Programm wird auf der Zentrale ausgeführt)

**Reply by SukramJ** [reply]
> Hmm, ist in der Tat nur bedingt aussagekräftig.

.. aber hilfreich. Das letzte Programm wird nicht automatisch entfernt.

**Reply by SukramJ** [reply]
Bei den Sysvars das Gleiche.

**Reply by SukramJ** [reply]
Weiter geht es mit der 1.76.0b3.

**Reply by Baxxy13** [reply]
Das sieht hier nun alles sehr gut aus und ich kann erstmal keine Probleme mehr erkennen.

Dieses "on-the-fly" anlegen/entfernen ist echt genial.

Super umgesetzt, danke Dir!

**Reply by SukramJ** [reply]
Danke für die Rückmeldung.
Noch eine Frage zu Alarm und Servicemeldungen:

Die heissen bei mir mal ${sysVarAlarmMessages}/${sysVarServiceMessages} und mal Alarmmeldungen/Servicemeldungen.
Warum werden beim ersteren die Platzhalter nicht ersetzt, und kann man dieses Sysvars auch unabhängig vom Namen identifizieren?

**Reply by Baxxy13** [reply]
Da muss ich etwas ausholen.
Das ist als "Klarnamen-Bug" im Forum bekannt.
Auf einem frischen System (egal ob CCU oder RM) sind die "echten Namen" `${sysVarAlarmMessages}` und `${sysVarServiceMessages}`
Da beide `systemintern` sind bekommt man die erstmal nicht zu Gesicht. 
Macht man sie "sichtbar" bekommt man als Name, je nach Spracheinstellung in der WebUI, die Übersetzung angezeigt.

Bei RM kann/muss man auf einem frischen System die Sprache wählen. Diese Funktion sorgt dafür das z.B. `${sysVarAlarmMessages}` zu `Alarmmeldungen` geändert wird.

Auf einer CCU gibt es das nicht, hier muss man die beiden SysVars aktiv selbst umbenennen. Was bekanntlich schwierig ist weil man generell erstmal ein Script bemühen muss um überhaupt dranzukommen.



> und kann man dieses Sysvars auch unabhängig vom Namen identifizieren?

Es gibt 2 Möglichkeiten.
A:
- beide haben eine fixe ID. Es wurde mal gemunkelt das das nicht immer so sein muss, habe ich aber noch die gesehen. 
- Alarmmeldungen = ID: 40 / Servicemeldungen = ID: 41

B:
- es gibt eine spezielle Zugriffsmethode die ohne ID funktioniert
- Alarmmeldungen = `dom.GetObject(ID_GW_SYSALARM);`
- Servicemeldungen = `dom.GetObject(ID_GW_SYSSERVICE);`

**Comment by SukramJ** [comment]
Apropos breaking change:

Mit Version 1.76.0 wird, wie angekündigt, auch das unignore File rausfliegen. Ich hoffe Du bist bereits umgezogen.

**Reply by Baxxy13** [reply]
Jetzt wo du es sagst schaue ich lieber nochmal durch. 
Glaube meine "nicht HA-Testsysteme" bauen noch auf das unignore-file.

**Reply by Baxxy13** [reply]
Tatsächlich setzten beide Haupt-HA's noch auf das unignore-file. 
Ist jetzt geändert. Danke für die Erinnerung. 😉
